### DISCUSSION TITLE
How can I use it with port extender?

### DISCUSSION THREAD

**Discussion by pruwait** [bxparks/AceButton#86] [General]
I want doble click and hold event on pcf8575.
How can I do it?

**Comment by bxparks** [comment]
Maybe? I am not familiar with the PCF8575. You will need to verify whether it supports reading its port every 5 milliseconds or so to detect the button presses. The datasheet suggests that it should, but it would be best to double check before doing too much work.

Then you will have to write a custom subclass of `ButtonConfig`  that translates the reading of the PCF8575 ports into something that looks like a `digitalRead()`. You can use this Discussion to ask me any followup questions about the AceButton framework if you get stuck somewhere.

[Edit: typo]

**Reply by pruwait** [reply]
esp32 + pcf8575 = every 4 milliseconds

```
20:49:10.175 -> READ VALUE FROM PCF in1: 	1848	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1852	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1856	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1860	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1865	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1869	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1873	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1877	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1882	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1886	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1890	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1894	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1899	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1903	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1907	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1911	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1916	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1920	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1924	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1928	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1933	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1937	1111111111111111
20:49:10.175 -> READ VALUE FROM PCF in1: 	1941	1111111111111111
```

I use Rob Tillaart [libruary  ](https://github.com/RobTillaart/PCF8575) and it has readButton16(), readButton16(mask), readButton(pin).
May be we can using this classes?

**Reply by bxparks** [reply]
Sure give it a try. A single I2C read transaction is reading all 16 lines. So you will want to process all buttons attached to the PCF8575 in a single batch, instead of making a separate I2C read transaction for each button. To do that, you will want to do something similar to my `EncodedButtonConfig` class, in particular, the `checkButtons()` method. 

In your `Pcf8575ButtonConfig::checkButtons()` method, you will loop through the 16 bits (or however many buttons you have attached), assign the virtual pin number for the bit (i.e. bit 0 corresponds to virtual pin 0), then call `AceButton::checkState()` for that button.

There are other ways of multiplexing multiple buttons on a small number of pins, but if you are constrained to use a PCF8575, then you will have to write this new `Pcf8575ButtonConfig` class. Just to be clear, you should not be waiting for me to do this work for you. I don't have access to a PCF8575 chip, I have other ways of getting multiple buttons on a few GPIO pins, and I don't have the motivation to spend time on this.
